_posts/2018-02-06-unreliable-network.md

```diff
@@ -0,0 +1,116 @@
+
+The idea with retrying requests is the following: when our communication API is called to send a request to the server, it does this normally and returns the successful response to the caller. However, if the request fails, our API waits a few seconds, and tries to send the request again. If it fails again, it waits longer, and tries again. If the request fails too many times, the API returns the failed request to the caller. In most cases, hopefully, the request is going to be successful after one or two attempts. The caller of the API does not know of the failed requests. It does not need to, as our API automativally keeps (re)sending the requests. The caller is informed about the failed request only if it has failed multiple times. Here is a naive example of a communication API written in [ClojureScript](https://clojurescript.org):
+
+```clojure
+(def default-retry-options {:timeout nil
+                             :attempts 5})
+
+(defn network-request
+  ([response-chan request-options]
+    ;; network-request called without retry-options, use the default value.
+   (network-request response-chan request-options default-retry-options))
+  ([response-chan request-options retry-options]
+   (let [response-handler (fn [[_ response]]
+                            (if (some? response)
+                              ;; Successful response, return it to the caller by using
+                              ;; the given response channel.
+                              (do (put! response-chan response)
+                                  (close! response-chan))
+                              ;; Failing response, any attempts left?
+                              (if (> (:attempts retry-options) 1)
+                                ;; Try again, decrease attempts by one and increase timeout
+                                (network-request
+                                  response-chan
+                                  request-options
+                                  {:timeout (+ (:timeout retry-options) 2000)
+                                   :attempts (dec (:attempts retry-options))})
+                                ;; Return the failing response to the caller
+                                (do (put! response-chan response)
+                                    (close! response-chan)))))]
+     ;; Send the network request, return the response to the response-handler
+     (go
+       ;; Before sending, wait for the timeout (if given)
+       (when-let [retry-timeout (:timeout retry-options)]
+         (<! (timeout retry-timeout)))
+       (ajax-request (merge
+                       request-options
+                       {:handler response-handler}))))))
+```
+
+In most cases, however, it is not practical to retry every single failed request. For example, if a request pointed to a file which was not found and returned 404, it probably does not help to try this request again. Thus, you should choose which requests you want to retry.
```

Comment:
The past tense is slightly odd here, so maybe "pointed" -> "points", "returned" -> "returns", etc., but it's not a big deal.